### DISCUSSION TITLE
How to make Philips hue gradient lightstrip show multiple colours

### DISCUSSION THREAD

**Discussion by gibahjoe** [awawa-dev/HyperHDR#469] [Q&A]
Hi, So I have Philips hue gradient lightstrip at the back of my tv. When I use Hyperhdr, it seems it only supports showing one colour on the gradient lightstrip. How can I fix this or work around it? Is there a setting I am missing?

**Comment by awawa-dev** [comment]
Hi
Gradient light-strip is only limited supported. To have it fully supported somebody must present or lend me such setup or request ready to use PR which implements it fully, as my previous experience with Philips Hue devices discourages to program them blindly.

**Reply by gibahjoe** [reply]
I could try to make a pr for this. Could you point me in the right direction on what to look out for?

**Reply by awawa-dev** [reply]
You need Philips developer account for the docs https://developers.meethue.com/ Also here, on github there are some implementations of the gradient lightstrips support that could be helpful if something in the docs is not clear. I can help you integrate your changes with HyperHDR if you have questions.

**Comment by gibahjoe** [comment]
Hi @awawa-dev ,

Merry Xmas.

I have sent a PR for this. Please review. Thanks
https://github.com/awawa-dev/HyperHDR/pull/471

**Reply by awawa-dev** [reply]
Thank you. Although duplicate with some modification almost 1.5k lines of code, makes it almost impossible to verify it especially when it comes to the security for the network device. Could the change be applied to LedDevicePhilipsHue.cpp with some conditional behavior for API v2? Is there any particular requirement for api v2 to change groupid from integer to string?

**Reply by gibahjoe** [reply]
In V2, groups IDs are strings (uuid). Hence the change.

I tried applying the changes conditionally on v1 but that was very messy because a lot has changed between v1 and v2.

For instance, all ids are now uuid strings, v2 uses channels which are very different from lights e.g, you can't turn off and on a channel. Also a channel can have multiple lights and a light can have multiple channels and there is no direct link between them.
There really is no duplicate code there as like I said, the apis are very very different which is why I couldn't conditionally apply them to v1.

**Reply by awawa-dev** [reply]
I understand, but changing this value type (group id) is a breaking change for old v1 API e.g. https://github.com/awawa-dev/HyperHDR/blob/27c42c69912153e43207a3a31a1e37cd5e3b3bc1/sources/leddevice/dev_net/LedDevicePhilipsHue.cpp#L1014. Probably a better solution is to add new field for it and make it conditional in the scheme (old field only for old API, new field for API v2, only one can be active at the time)

**Reply by awawa-dev** [reply]
I think this will fail on ipv6:
![obraz](https://user-images.githubusercontent.com/69086569/209816159-75993205-a77f-421b-9891-95f9da253376.png)

**Reply by gibahjoe** [reply]
hi, I have separated the variables and also fixed this possible issue with ipv6

**Reply by awawa-dev** [reply]
Great! I will test it later with my setup.

**Reply by awawa-dev** [reply]
1 Hue wizard (v1) doesnt work anymore
![obraz](https://user-images.githubusercontent.com/69086569/209876615-e169b196-7d18-4ed8-afc6-b9fc3a80cddc.png)
2 Does  httpResponse::QMap<QString, QString> getHeaders() const; (and headers) are needed for something? there is no reference for them in the app.

**Reply by gibahjoe** [reply]
1) resolved
2). Yes headers are used to get the application ID which determines which application is currently streaming to the Lights. You can find usage here in the PR
<img width="965" alt="Screenshot 2022-12-28 at 23 13 16" src="https://user-images.githubusercontent.com/9147147/209879228-7f981711-cf56-4340-a98d-1a0c2f88bdc9.png">

**Reply by awawa-dev** [reply]
Thank you for the prompt response. Can confirm the wizard is working again. My VS2022 didn't refresh new files in leddevice folder so that's why it couldn't find the reference. Everything is fine now.
